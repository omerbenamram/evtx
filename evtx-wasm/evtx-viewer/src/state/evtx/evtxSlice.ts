import type { EvtxFileInfo } from "../../lib/types";

// ---------------- Types & State ----------------
export interface EvtxMetaState {
  isLoading: boolean;
  loadingMessage: string;
  matchedCount: number;
  totalRecords: number;
  fileInfo: EvtxFileInfo | null;
  currentFileId: string | null;
}

export const evtxInitialState: EvtxMetaState = {
  isLoading: false,
  loadingMessage: "",
  matchedCount: 0,
  totalRecords: 0,
  fileInfo: null,
  currentFileId: null,
};

// ---------------- Action Types ----------------
export type EvtxAction = {
  type: "evtx/UPDATE";
  payload: Partial<EvtxMetaState>;
};

// ---------------- Reducer ----------------
export function evtxReducer(
  state: EvtxMetaState = evtxInitialState,
  action: EvtxAction
): EvtxMetaState {
  switch (action.type) {
    case "evtx/UPDATE":
      return { ...state, ...action.payload };
    default:
      return state;
  }
}

// ---------------- Action Creators ----------------
export const updateEvtxMeta = (
  payload: Partial<EvtxMetaState>
): EvtxAction => ({
  type: "evtx/UPDATE",
  payload,
});
