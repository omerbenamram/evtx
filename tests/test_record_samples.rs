mod fixtures;
use fixtures::*;

use evtx::{EvtxParser, ParserSettings};
use serde_json::Value;

#[test]
fn test_event_xml_sample() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/security.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    insta::assert_display_snapshot!(first_record.data);
}

#[test]
fn test_event_json_sample() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/security.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records_json()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&first_record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}

#[test]
fn test_event_json_sample_with_event_data() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/2-system-Security-dirty.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records_json()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&first_record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}

#[test]
fn test_event_xml_sample_with_event_data() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/2-system-Security-dirty.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    insta::assert_display_snapshot!(first_record.data);
}

#[test]
fn test_event_json_sample_with_event_data_with_attributes_and_text() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/system.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records_json()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&first_record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}

#[test]
fn test_event_xml_sample_with_event_data_with_attributes_and_text() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/system.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    insta::assert_display_snapshot!(first_record.data);
}

#[test]
fn test_event_xml_sample_with_user_data() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!(
        "../samples/E_Windows_system32_winevt_logs_Microsoft-Windows-CAPI2%4Operational.evtx"
    );
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record = parser
        .records()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    insta::assert_display_snapshot!(first_record.data);
}

#[test]
fn test_event_xml_sample_with_entity_ref() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!(
        "../samples/E_Windows_system32_winevt_logs_Microsoft-Windows-CAPI2%4Operational.evtx"
    );
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let record = parser
        .records()
        .filter_map(|record| record.ok())
        .find(|record| record.event_record_id == 28)
        .expect("record to parse correctly");

    insta::assert_display_snapshot!(record.data);
}

#[test]
fn test_event_xml_sample_with_entity_ref_2() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!(
        "../samples/E_Windows_system32_winevt_logs_Microsoft-Windows-CAPI2%4Operational.evtx"
    );
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let record = parser
        .records()
        .filter_map(|record| record.ok())
        .find(|record| record.event_record_id == 25)
        .expect("record to parse correctly");

    insta::assert_display_snapshot!(record.data);
}

#[test]
fn test_event_json_with_multiple_nodes_same_name() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!(
        "../samples/E_Windows_system32_winevt_logs_Microsoft-Windows-CAPI2%4Operational.evtx"
    );
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let record = parser
        .records_json()
        .filter_map(|record| record.ok())
        .find(|record| record.event_record_id == 28)
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}

#[test]
fn test_event_json_sample_with_separate_json_attributes() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/Application.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(
            ParserSettings::new()
                .num_threads(1)
                .separate_json_attributes(true),
        );

    let first_record = parser
        .records_json()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&first_record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}

#[test]
fn test_event_json_with_multiple_data_elements() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/MSExchange_Management_wec.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(ParserSettings::new().num_threads(1));

    let first_record_xml = parser
        .records()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    insta::assert_display_snapshot!(&first_record_xml.data);

    let first_record = parser
        .records_json()
        .next()
        .expect("to have records")
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&first_record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}

#[test]
fn test_event_json_with_multiple_nodes_same_name_separate() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!(
        "../samples/E_Windows_system32_winevt_logs_Microsoft-Windows-CAPI2%4Operational.evtx"
    );
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(
            ParserSettings::new()
                .num_threads(1)
                .separate_json_attributes(true),
        );

    let record = parser
        .records_json()
        .filter_map(|record| record.ok())
        .find(|record| record.event_record_id == 28)
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}

#[test]
fn test_event_json_multiple_empty_data_nodes_not_ignored() {
    ensure_env_logger_initialized();
    let evtx_file = include_bytes!("../samples/issue_201.evtx");
    let mut parser = EvtxParser::from_buffer(evtx_file.to_vec())
        .unwrap()
        .with_configuration(
            ParserSettings::new()
                .num_threads(1)
                .separate_json_attributes(true),
        );

    let record = parser
        .records_json()
        .filter_map(|record| record.ok())
        .next()
        .expect("record to parse correctly");

    let value: Value = serde_json::from_str(&record.data).expect("to parse correctly");
    insta::assert_json_snapshot!(&value);
}
