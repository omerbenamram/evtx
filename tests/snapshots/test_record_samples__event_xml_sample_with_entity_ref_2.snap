---
source: tests/test_record_samples.rs
assertion_line: 166
expression: record.data
---
<?xml version="1.0" encoding="utf-8"?>
<Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
  <System>
    <Provider Name="Microsoft-Windows-CAPI2" Guid="{5bbca4a8-b209-48dc-a8c7-b23d3e5216fb}">
    </Provider>
    <EventID>51</EventID>
    <Version>0</Version>
    <Level>4</Level>
    <Task>51</Task>
    <Opcode>2</Opcode>
    <Keywords>0x400000000000002e</Keywords>
    <TimeCreated SystemTime="2017-05-19T02:02:36.187500Z">
    </TimeCreated>
    <EventRecordID>25</EventRecordID>
    <Correlation>
    </Correlation>
    <Execution ProcessID="1396" ThreadID="2132">
    </Execution>
    <Channel>Microsoft-Windows-CAPI2/Operational</Channel>
    <Computer>WIN-M5327EF98B9</Computer>
    <Security UserID="S-1-5-21-1223297778-3299746493-1462173606-500">
    </Security>
  </System>
  <UserData>
    <CryptRetrieveObjectByUrlCache>
      <URL scheme="http">http://crl.microsoft.com/pki/crl/products/microsoftrootcert.crl</URL>
      <Object type="CONTEXT_OID_CRL" constant="2">
      </Object>
      <Flags value="2003" CRYPT_RETRIEVE_MULTIPLE_OBJECTS="true" CRYPT_CACHE_ONLY_RETRIEVAL="true" CRYPT_LDAP_SCOPE_BASE_ONLY_RETRIEVAL="true">
      </Flags>
      <AuxInfo maxUrlRetrievalByteCount="104857600">
      </AuxInfo>
      <CacheInfo lastSyncTime="2017-02-03T20:08:47.908Z">
        <URLCachePrefetchInfo objectType="CRYPTNET_URL_CACHE_PRE_FETCH_CRL" thisUpdateTime="2016-11-20T21:48:50Z" nextUpdateTime="2017-02-19T10:08:50Z" publishTime="2017-02-18T21:58:50Z">
        </URLCachePrefetchInfo>
        <URLCacheFlushInfo expireTime="2017-02-19T10:08:50Z">
        </URLCacheFlushInfo>&quot;&quot;<URLCacheResponseInfo responseType="CRYPTNET_URL_CACHE_RESPONSE_HTTP" responseValidated="true" lastModifiedTime="2016-11-21T06:01:26Z" maxAge="900" eTag="ea9ee7b1bc43d21:0">
        </URLCacheResponseInfo>
      </CacheInfo>
      <RetrievedObjects>
        <CertificateRevocationList fileRef="B2A57B544F9B889695FC28CFE5EEDAA85B4B6602.crl" issuerName="Microsoft Root Certificate Authority">
        </CertificateRevocationList>
      </RetrievedObjects>
      <EventAuxInfo ProcessName="Setup.exe">
      </EventAuxInfo>
      <CorrelationAuxInfo TaskId="{74E4CD40-C966-49F5-B50A-032DFFEE57CA}" SeqNumber="4">
      </CorrelationAuxInfo>
      <Result value="0">
      </Result>
    </CryptRetrieveObjectByUrlCache>
  </UserData>
</Event>
